aer/src/runtime/cell_service.rs

```diff
@@ -0,0 +1,151 @@
+/* -------------------------------------------------------------------------- *\
+ *          Apache 2.0 License Copyright © 2022-2023 The Aurae Authors        *
+ *                                                                            *
+ *                +--------------------------------------------+              *
+ *                |   █████╗ ██╗   ██╗██████╗  █████╗ ███████╗ |              *
+ *                |  ██╔══██╗██║   ██║██╔══██╗██╔══██╗██╔════╝ |              *
+ *                |  ███████║██║   ██║██████╔╝███████║█████╗   |              *
+ *                |  ██╔══██║██║   ██║██╔══██╗██╔══██║██╔══╝   |              *
+ *                |  ██║  ██║╚██████╔╝██║  ██║██║  ██║███████╗ |              *
+ *                |  ╚═╝  ╚═╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═╝╚══════╝ |              *
+ *                +--------------------------------------------+              *
+ *                                                                            *
+ *                         Distributed Systems Runtime                        *
+ *                                                                            *
+ * -------------------------------------------------------------------------- *
+ *                                                                            *
+ *   Licensed under the Apache License, Version 2.0 (the "License");          *
+ *   you may not use this file except in compliance with the License.         *
+ *   You may obtain a copy of the License at                                  *
+ *                                                                            *
+ *       http://www.apache.org/licenses/LICENSE-2.0                           *
+ *                                                                            *
+ *   Unless required by applicable law or agreed to in writing, software      *
+ *   distributed under the License is distributed on an "AS IS" BASIS,        *
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ *   See the License for the specific language governing permissions and      *
+ *   limitations under the License.                                           *
+ *                                                                            *
+\* -------------------------------------------------------------------------- */
+
+use crate::execute;
+use aurae_client::runtime::cell_service::CellServiceClient;
+use aurae_proto::runtime::{
+    Cell, CellServiceAllocateRequest, CellServiceFreeRequest,
+    CellServiceStartRequest, CellServiceStopRequest, CpuController,
+    CpusetController, Executable,
+};
+use clap::Subcommand;
+
+// We'd like to macro this struct and the impl from the CellService in the proto.
+// I'm (future-highway) not sure how, even though this seems almost entirely boilerplate.
+//
+// Maybe macros is the wrong approach.
```

Comment:
Hrmmm I see what you mean here. Having to define the `CellServiceCommands` enum for each command seems like a lot of work. This feels similar to what I originally was trying to do with AuraeScript before it become much more beautiful like it is now.

Why won't macros work exactly? This enum looks like something we should be able to generate with a macro?

Reply:
For proc_macros, which are the macros that are in their own crates (as opposed to `macro_rules`) we have 3 choices of macros. Two use attributes (e.g., `#[derive(MyMacro)]` and `#[my_macro]`), and need a struct or enum for us to "attach" them. We can overcome this hurdle by generating the rust code using build.rs like we did prior to buf, but then we would be left with attaching attributes in the build.rs, which is also not great.

The other option is to make a macro like those used in `auraescript` and `aurae-client` where we essentially define our own syntax. That may make the code shorter, but not necessarily better to deal with (we'd need to update the macro each time we have a feature of `clap` that isn't made available in the macro syntax yet).

If we had a proto parser, then we may be able to significantly reduce the code needed and have a better dev experience, but I was unable to find a rust crate that seemed suitable.

Reply:
Well, now we have a proto parser in use for aurae-client and soon auraescript.

I think the creation of macro of aer to use it will be more work then the conversion of the other crates, but it should be technically possible. It won't be as minimal as we still want to be able to use clap attributes I would think, but it should be possible to significantly reduce the code needed.